pkg/machine/hyperv/vsock.go

```diff
@@ -0,0 +1,267 @@
+	if err != nil {
+		return err
+	}
+	if exists {
+		return fmt.Errorf("%q: %s", ErrVSockRegistryEntryExists, hv.KeyName)
+	}
+	parentKey, err := registry.OpenKey(registry.LOCAL_MACHINE, VSOCK_REGISTRY_PATH, registry.QUERY_VALUE)
+	defer func() {
+		if err := parentKey.Close(); err != nil {
+			logrus.Error(err)
+		}
+	}()
+	if err != nil {
+		return err
+	}
+	newKey, _, err := registry.CreateKey(parentKey, hv.KeyName, registry.WRITE)
+	defer func() {
+		if err := newKey.Close(); err != nil {
+			logrus.Error(err)
+		}
+	}()
+	if err != nil {
+		return err
+	}
+
+	if err := newKey.SetStringValue(HVSOCK_PURPOSE, hv.Purpose.string()); err != nil {
+		return err
+	}
+	return newKey.SetStringValue(HVSOCK_MACHINE_NAME, hv.MachineName)
+}
+
+// Remove deletes the registry key and its string values
+func (hv *HVSockRegistryEntry) Remove() error {
+	return registry.DeleteKey(registry.LOCAL_MACHINE, hv.fqPath())
+}
+
+func (hv *HVSockRegistryEntry) fqPath() string {
+	return fmt.Sprintf("%s\\%s", VSOCK_REGISTRY_PATH, hv.KeyName)
+}
+
+func (hv *HVSockRegistryEntry) validate() error {
+	if hv.Port < 1 {
+		return errors.New("port must be larger than 1")
+	}
+	if len(hv.Purpose.string()) < 1 {
+		return errors.New("required field purpose is empty")
+	}
+	if len(hv.MachineName) < 1 {
+		return errors.New("required field machinename is empty")
+	}
+	if len(hv.KeyName) < 1 {
+		return errors.New("required field keypath is empty")
+	}
+	//decimal_num, err = strconv.ParseInt(hexadecimal_num, 16, 64)
+	return nil
+}
+
+func (hv *HVSockRegistryEntry) exists() (bool, error) {
+	foo := hv.fqPath()
+	_ = foo
+	_, err := openVSockRegistryEntry(hv.fqPath())
+	if err == nil {
+		return true, err
+	}
+	if errors.Is(err, registry.ErrNotExist) {
+		return false, nil
+	}
+	return false, err
+}
+
+// findOpenHVSockPort looks for an open random port. it verifies the port is not
+// already being used by another hvsock in the Windows registry.
+func findOpenHVSockPort() (uint64, error) {
+	// If we cannot find a free port in 10 attempts, something is wrong
+	for i := 0; i < 10; i++ {
+		port, err := utils.GetRandomPort()
+		if err != nil {
+			return 0, err
+		}
+		// Try and load registry entries by port to see if they exist
+		_, err = LoadHVSockRegistryEntry(uint64(port))
+		if err == nil {
+			// the port is no good, it is being used; try again
+			logrus.Errorf("port %d is already used for hvsock", port)
+			continue
+		}
+		if errors.Is(err, registry.ErrNotExist) {
+			// the port is good to go
+			return uint64(port), nil
+		}
+		if err != nil {
+			// something went wrong
+			return 0, err
+		}
+	}
+	return 0, errors.New("unable to find a free port for hvsock use")
+}
+
+// NewHVSockRegistryEntry is a constructor to make a new registry entry in Windows.  After making the new
+// object, you must call the add() method to *actually* add it to the Windows registry.
+func NewHVSockRegistryEntry(machineName string, purpose HVSockPurpose) (*HVSockRegistryEntry, error) {
```

Comment:
Will this function be called outside this package in the future? If not, we can make it private and put it in lower caps.